Function Naming

lrennels opened the pull request:
- remove unnecessary terms from `export` list
- apply changes to the naming, following conventions for `!`, `_`, and naming of `component`, `parameter`, and `variable`
- follow through on these changes throughout `Mimi` code base, documentation (including `Porting` doc), and other models

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
-    compdefs,
-    compinstance,
-    compkeys,
+    addcomponent,  # push!, add_component!, add_component, addcomp!, add!; could also have multiple functions like insert!, push! etc. instead of before after keywords
```
suggestion:  `add_comp`
reason:  matches `_` convention; matches using `comp, var, and param`; `add!` seems perhaps too simplified for now since we generally do specific comp, var, or param

rjplevin replied:
I agree.

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
-    compinstance,
-    compkeys,
+    addcomponent,  # push!, add_component!, add_component, addcomp!, add!; could also have multiple functions like insert!, push! etc. instead of before after keywords
+    add_connector_comps, #should this be deleted?  Docs indicate we need it ...
```
This gets automatically called within each `@defcomp`, but I think we still need it at the end of the API for building models.

rjplevin replied:
We could make `add_connector_comps` the first step of `build(m::Model)`. This would allow us to remove it from `@defcomp` and alleviate the user from calling it interactively. Any objections?

lrennels replied:
I like that!  I'm not fully comfortable with the internals of editing a built model yet ... but I assume any time a big change is made which requires a new call to `add_connector_comps`, we probably `decache` and then `build` again.  That question doesn't affect this decision, just trying to be clear when I'm editing docs.

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
+    add_connector_comps, #should this be deleted?  Docs indicate we need it ...
     components,
-    connect_parameter,
+    connect_parameter, # connect!, bind, assign, link; ! yes or no; frontrunner: connect!
```
suggestion:  `connect_param`
reason:  matches the convention for `!` because this doesn't change something that exists (unless you mean the model itself but that's not generally what we mean with `!`); matches convention of `<action>_<comp/param/var>` that we generally seem to follow

rjplevin replied:
Actually, I'd argue for `!`. I interpret that as indicating a function with side-effects, i.e., not a pure function that returns a value but leaves all else unchanged.

lrennels replied:
Ok that might be a more consistent interpretation, maybe above should be `add_comp!` then?

rjplevin replied:
Yes.

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
-    dimension,
-    dimensions,
-    disconnect!,
+    disconnect!, # disconnect_parameter!; frontrunner: disconnect!
```
suggestion:  `disconnect_param!`
reason:  matches the above `connect_param` but uses `!` since it alters an existing connection

rjplevin replied:
Agreed.

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
-    setproperty!,
-    unit,
-    variables
+    replace_component, # Add !, comp vs, component, replace!
```
suggestion:  `replace_comp!`
reason:  matches `!` convention;  matches aforementioned `<action>_<comp/param/var>` pattern

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
-    unit,
-    variables
+    replace_component, # Add !, comp vs, component, replace!
+    set_dimension!, # Think hard about axis-dimension-index-blabla
```
leave this alone for now since this is a bigger design discussion?

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
-    variables
+    replace_component, # Add !, comp vs, component, replace!
+    set_dimension!, # Think hard about axis-dimension-index-blabla
+    set_leftover_params!, #rethink in general
```
leave this alone for now, we may want to discuss the use cases of this function but I think that's a bigger discussion for later

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
+    replace_component, # Add !, comp vs, component, replace!
+    set_dimension!, # Think hard about axis-dimension-index-blabla
+    set_leftover_params!, #rethink in general
+    set_parameter!, # Just make parameter match other parts
```
suggestion:  `set_param!`
reason:  matches other parts

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
-    getproperty,
     gettime,
     get_parameter_value,
     get_variable_value,
```
`get_parameter_value ` -->  `get_param_value` to match `parameter` --> `param`?
`get_variable_value ` -->  `get_var_value` to match `variable` --> `var`?

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
     load_comps,
     modeldef,
     name,
     new_component,
```
`new_component ` -->  `new_comp` to match `component` --> `comp`?

rjplevin replied:
Agreed.

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
-    run_timestep,
-    set_dimension!,
-    set_leftover_params!,
-    set_parameter!,
```
Change to `set_param!` event though this will be removed from the API?

rjplevin replied:
Agreed.

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
     modeldef,
     name,
     new_component,
     parameters,
```
`parameters ` -->  `params` to match `parameter` --> `param`?

lrennels commented on src/Mimi.jl:
```diff
@@ -14,50 +14,32 @@ using StringBuilders
     parameters,
-    # plot,
-    # plot_comp_graph,
-    replace_component,
```
`replace_component ` -->  `replace_comp` to match `component` --> `comp`?

rjplevin replied:
Agreed.

lrennels commented:
@davidanthoff @rjplevin This PR is ready, I have done all the naming changes here, updated the documentation, and updated integration branches.  There are four "TODO" comments in the export list, which I will handle in my next PR since they are slightly more involved and I'd rather keep them separate.
